### ISSUE TITLE
User-defined literals does not work, even with -Wno-error=reserved-user-defined-literal enabled

### ISSUE THREAD

**Issue by tada123** [llvm/llvm-project#76394]
Trying to compile following code:
```cpp
#include <cstddef>
#include <cstdlib>
#include <cstring>
#include <cstdio>

class MyString {
public:
    size_t sz;
    char* str;
    
    inline ~MyString(){
        free(this->str);
    }
    inline char* c_str(){
        return str;
    }
};

MyString operator "" M(const char* str, size_t sz){
    return MyString {sz, strdup(str)};
}

MyString test(){
    MyString s = "Qqqq"M;
    return s;
}
int main(int argc, char** argv){
    puts(test().c_str());
}
```
Compiling with only `g++` works fine with this output:
```
clang_literal.cpp:20:10: warning: literal operator suffixes not preceded by ‘_’ are reserved for future standardization [-Wliteral-suffix]
   20 | MyString operator "" M(const char* str, size_t sz){
      |          ^~~~~~~~
```
However, compilation with `clang++` gives error messages even with `-Wno-error=reserved-user-defined-literal enabled` enabled:

```
 >>>  clang++ -Wno-error=reserved-user-defined-literal -Wno-reserved-user-defined-literal clang_literal.cpp
clang_literal.cpp:20:10: warning: user-defined literal suffixes not starting with '_' are reserved; no literal will invoke this operator [-Wuser-defined-literals]
MyString operator "" M(const char* str, size_t sz){
         ^
clang_literal.cpp:25:14: error: no viable conversion from 'const char[5]' to 'MyString'
    MyString s = "Qqqq"M;
             ^   ~~~~~~
clang_literal.cpp:7:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char[5]' to 'const MyString &' for 1st argument
class MyString {
      ^
clang_literal.cpp:25:24: error: expected ';' at end of declaration
    MyString s = "Qqqq"M;
                       ^
                       ;
1 warning and 2 errors generated.
```
Yes, i know, that adding underscore fixes the problem, but i have really huge project, which does not use underscores for custom literals.
Is it a bug, or is there some another flag/workaround (except `-std=c++9`), that must be enabled to allow the code to compile with `clang++` same way as `g++` does?

**Comment by shafik** [comment]
CC @cor3ntin

**Comment by eisenwave** [comment]
Kinda annoying. Ran into this in
- https://github.com/eisenwave/std-big-int/pull/34

Implement UDLs for `std::big_int` doesn't really work when it's a hard error at the lexer level, even when a UDL exists. I get that it should warn when you do this, but it's making it literally impossible to create reference implementations of standard features that demonstrate a proposed UDL.

**Comment by zygoloid** [comment]
The problem is that *ud-suffix*es for string literals were a big breaking change, for example for code that made idiomatic use of the `PRI*` and `SCN*` macros from `<inttypes.h>`:

```c++
void f(uint32_t n) {
  printf("%"PRIu32, n);
}
```

In order to break as little code as possible, Clang treats a string literal followed by an identifier as a *ud-suffix* only if the suffix is a standard-defined suffix or a suffix that users are allowed to define.

(I believe GCC uses a different approach here, where they check whether the *ud-suffix* is defined as a macro and either lex as one token or as two on that basis. That would require an unpleasant amount of feedback from the preprocessor into the lexer, and seems more brittle than Clang's approach.)

**Comment by zygoloid** [comment]
> That would require an unpleasant amount of feedback from the preprocessor into the lexer, and seems more brittle than Clang's approach.

Maybe we can avoid the feedback by implementing this in the preprocessor rather than the lexer: we could use the formal rule for lexing (so we always produce a user-defined string literal when a string literal is followed by a ud-suffix), but if preprocessing encounters a user-defined string literal token, it could check to see if the suffix is a defined macro and split the token into two if needed. That might be a reasonable path forward, and would presumably improve GCC compatibility.

**Comment by eisenwave** [comment]
> In order to break as little code as possible, Clang treats a string literal followed by an identifier as a ud-suffix only if the suffix is a standard-defined suffix or a suffix that users are allowed to define.

What about integer literals? Is that a separate issue then? I've tried to make `123n` work for `std::big_int` and that also got rejected as a hard error, and I don't see any breaking change caused by that.

**Comment by zygoloid** [comment]
> > In order to break as little code as possible, Clang treats a string literal followed by an identifier as a ud-suffix only if the suffix is a standard-defined suffix or a suffix that users are allowed to define.
> 
> What about integer literals? Is that a separate issue then? 

Yes, the concerns there are quite different. _Lexing_ isn't a problem there, since the shape of the token is the same regardless. We don't need to know whether the characters on the end are a ud-suffix or something else until we're in Sema, where we can do an actual lookup.

However, lots of letters as suffixes are already in use -- `bfhijklrquwz` are all supported as suffixes (or parts of suffixes) in clang, and most of them are allowed in any case and in any order. This obviously creates a problem allowing user-defined suffixes that use those letters. Even the standard library suffixes aren't immune -- the suffixes `i`, `if`, and `il` conflict with Clang extensions that predate the introduction of those suffixes into the C++ standard library. What we do for those is to first look up the suffix as a literal operator, and then if it doesn't exist, we fall back to treating it as a builtin suffix. But we only do that for those three suffixes, not for all suffixes, since it would be very expensive in some cases (for example, very large tables of literals in generated code) to do a literal operator name lookup for all suffixes (eg, for every `UL` suffix).

We could pick a different tradeoff there, and probably should -- I think our behavior there is more conservative than it really needs to be. It'd probably be reasonable to detect whether a suffix is a "standard" suffix (that is, standard in the current language mode rather than as an extension), and do literal operator lookup for all non-standard suffixes (falling back to extension behavior if the literal operator is not declared) rather than only doing literal operator lookup for non-reserved and library-defined suffixes (and only doing the fallback in the intersection of library-defined + language extension suffixes).
